injector/network_disruption.go

```diff
@@ -374,78 +416,442 @@ func (i *networkDisruptionInjector) applyOperations() error {
 	return nil
 }
 
-// getServices parses the Kubernetes services in the disruption spec and returns a set of (ip, port, protocol) tuples
-func (i *networkDisruptionInjector) getServices() ([]networkDisruptionService, error) {
-	allServices := []networkDisruptionService{}
+// addServiceFilter add a tc service filter on a list of interfaces
+func (i *networkDisruptionInjector) addServiceFilter(service networkDisruptionService, interfaces []string, flowid string) (map[string][]string, error) {
+	computedPreferencesPerIface := make(map[string][]string)
 
-	for _, serviceSpec := range i.spec.Services {
-		// retrieve serviceSpec
-		k8sService, err := i.config.K8sClient.CoreV1().Services(serviceSpec.Namespace).Get(context.Background(), serviceSpec.Name, metav1.GetOptions{})
-		if err != nil {
-			return nil, fmt.Errorf("error getting the given kubernetes service (%s/%s): %w", serviceSpec.Namespace, serviceSpec.Name, err)
+	// We list filters before and after to find the new filters' preferences created. Those preferences are used to later remove those filters
+	filtersBeforeAddition, err := i.config.TrafficController.ListFilters(interfaces)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't identify tc filter created, error on tc filter listing: %v", err.Error())
+	}
+
+	err = i.config.TrafficController.AddFilter(interfaces, "1:0", 0, nil, service.ip, 0, service.port, service.protocol, flowid)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't create tc filter: %v", err.Error())
+	}
+
+	filtersAfterAddition, err := i.config.TrafficController.ListFilters(interfaces)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't identify tc filter created, error on tc filter listing: %v", err.Error())
+	}
+
+	for _, iface := range interfaces {
```

Comment:
Any chance you could break the logic in this for loop out into a new function that gets unit tested? It seems potentially brittle

Reply:
I changed the logic to keep track of the priority in another way: we can directly add a filter with a priority.